gemrb/core/Map.cpp

```diff
@@ -2793,246 +2796,203 @@ PathNode* Map::GetLine(const Point &start, const Point &dest, int Speed, int Ori
+	static const char dx[DEGREES_OF_FREEDOM] = {1, 0, -1, 0, 1, 1, -1, -1};
+	static const char dy[DEGREES_OF_FREEDOM] = {0, 1, 0, -1, 1, -1, 1, -1};
+	static const float weight = 1.5;	// Weighted heuristic. Finds sub-optimal paths
+										// but should be quite a bit faster
+	SearchmapPoint smptSource;
+	SearchmapPoint smptDest;
+	smptSource.x = s.x / 16;
+	smptSource.y = s.y / 12;
+	smptDest.x = d.x / 16;
+	smptDest.y = d.y / 12;
+	// Initialize data structures
+	//std::priority_queue<PQNode, std::vector<PQNode>, std::greater<PQNode>> open;
+	FibonacciHeap<PQNode> open;
+	bool *isClosed = new bool[Width * Height]();
+
+	SearchmapPoint *parents = new SearchmapPoint[Width * Height];
+	unsigned short *distFromStart = new unsigned short[Width * Height]();
+	int dxCross = smptDest.x - smptSource.x;
+	int dyCross = smptDest.y - smptSource.y;
+
+	if (size && GetBlocked(d.x, d.y, size)) {
+		AdjustPosition(smptDest);
+	}
+
+	Log(DEBUG, logString, "FindPath((%d %d), (%d %d), minDist = %d, size = %d)",
+		smptSource.x, smptSource.y, smptDest.x, smptDest.y, minDistance, size);
+
+	bool foundPath = false;
+	parents[smptSource.y * Width + smptSource.x] = smptSource;
+	open.emplace(PQNode(smptSource, 0));
+	unsigned int squaredMinDist = minDistance * minDistance;
+
+	// Avoid calling constructors by declaring object variables here
+	// Yes, there is an overhead in creating a Point a kajillion times
+	SearchmapPoint smptCurrent;
+	SearchmapPoint smptChild;
+	SearchmapPoint smptParent;
+	SearchmapPoint smptBestParent;
+	SearchmapPoint smptCurNeighbor;
+	NavmapPoint nmptCurrent;
+	NavmapPoint nmptAdjustedCurrent;
+	NavmapPoint nmptParent;
+	PQNode newNode;
+	while (!open.empty()) {
+
+		smptCurrent = open.top().first;
+		open.pop();
+
+		nmptCurrent.x = smptCurrent.x * 16;
+		nmptCurrent.y = smptCurrent.y * 12;
+		nmptAdjustedCurrent.x = nmptCurrent.x + 8;
+		nmptAdjustedCurrent.y = nmptCurrent.y + 6;
+
+
+		if (smptCurrent == smptDest) {
+			foundPath = true;
 			break;
-		} else if (MinDistance) {
-			/* check minimum distance:
-			 * as an obvious optimisation we only check squared distance: this is a
-			 * possible overestimate since the sqrt Distance() rounds down
-			 * (some other optimisations could be made here, but you'd be better off
-			 * fixing the pathfinder to do A* properly)
-			 * caller should have already done PersonalDistance adjustments, this is
-			 * simply between the specified points
-			 */
-
-			int distx = (x*16 + 8) - d.x;
-			int disty = (y*12 + 6) - d.y;
-			if ((unsigned int)(distx*distx + disty*disty) <= squaredmindistance) {
-				// we are within the minimum distance of the goal
-				Point ourpos(x*16 + 8, y*12 + 6);
-				// sight check is *slow* :(
-				if (!sight || IsVisibleLOS(ourpos, d)) {
-					// we got all the way to a suitable goal!
-					goal = Point(x, y);
-					found_path = true;
+		} else if (minDistance) {
+			int xDist = nmptAdjustedCurrent.x + 8 - d.x;
+			int yDist = nmptAdjustedCurrent.y + 6 - d.y;
```

Comment:
Interesting that another half-cell is added here, basically always looking at the diagonal neighbour of nmptCurrent.